File: SpectrumAnalyzer/WinMM/Wave.cs
using System;
using System.Runtime.InteropServices;
using System.Threading;

namespace WinMM {
	public abstract class Wave : IDisposable {
		public enum EAvailableFormats : uint {
			MONO_8bit_11kHz = 0x1,
			MONO_8bit_22kHz = 0x10,
			MONO_8bit_44kHz = 0x100,
			MONO_8bit_48kHz = 0x1000,
			MONO_8bit_96kHz = 0x10000,
			STEREO_8bit_11kHz = 0x2,
			STEREO_8bit_22kHz = 0x20,
			STEREO_8bit_44kHz = 0x200,
			STEREO_8bit_48kHz = 0x2000,
			STEREO_8bit_96kHz = 0x20000,
			MONO_16bit_11kHz = 0x4,
			MONO_16bit_22kHz = 0x40,
			MONO_16bit_44kHz = 0x400,
			MONO_16bit_48kHz = 0x4000,
			MONO_16bit_96kHz = 0x40000,
			STEREO_16bit_11kHz = 0x8,
			STEREO_16bit_22kHz = 0x80,
			STEREO_16bit_44kHz = 0x800,
			STEREO_16bit_48kHz = 0x8000,
			STEREO_16bit_96kHz = 0x80000
		}

		[Flags]
		protected enum WHDR_FLAGS : uint {
			NONE = 0,
			DONE = 0x00000001,
			PREPARED = 0x00000002,
			BEGINLOOP = 0x00000004,
			ENDLOOP = 0x00000008,
			INQUEUE = 0x00000010
		}

		[StructLayout(LayoutKind.Sequential)]
		protected struct WAVEFORMATEX {
			public ushort wFormatTag;
			public ushort nChannels;
			public uint nSamplesPerSec;
			public uint nAvgBytesPerSec;
			public ushort nBlockAlign;
			public ushort wBitsPerSample;
			public ushort cbSize;
		}

		[StructLayout(LayoutKind.Sequential)]
		protected struct WAVEHDR {
			public IntPtr lpData;
			public uint dwBufferLength;
			public uint dwBytesRecorded;
			public UIntPtr dwUser;
			public WHDR_FLAGS dwFlags;
			public uint dwLoops;
			public IntPtr lpNext;
			public UIntPtr reserved;
		}

		protected const uint DefaultDevice = unchecked((uint)-1);
		protected const int DeviceTimeout = 2000;
		protected const int BufferTaskTimeout = 500;

		protected IntPtr DeviceHandle;
		protected WAVEFORMATEX WaveFormatEx;
		protected readonly IntPtr[] WaveHeaderPtrs;
		protected readonly int BufferFrames;
		protected readonly int BufferSize;
		protected readonly int BufferCount;
		protected readonly byte[] MuteData;
		protected readonly AutoResetEvent DeviceOpened;
		protected readonly AutoResetEvent DeviceClosed;
		protected readonly AutoResetEvent CallbackEnabled;
		protected readonly AutoResetEvent BufferReady;
		protected volatile bool NotifyClose;
		protected volatile bool NotifyEndOfFile;
		protected volatile bool NotifyMute;
		protected volatile bool Muted;
		private IntPtr TotalWaveHeader;
		private IntPtr TotalBuffer;
		private Thread BufferThread;

		public const ushort FormatTag = 3;
		public const ushort Bits = 32;
		public const ushort Channels = 2;
		public readonly int SampleRate;

		public uint DeviceId { get; private set; }

		public bool IsPlaying => !Muted;

		protected Wave(int sampleRate, double unitTime, int unitCount) {
			var bytesPerFrame = Bits * Channels >> 3;
			SampleRate = sampleRate;
			DeviceId = DefaultDevice;
			BufferFrames = (int)(sampleRate * unitTime) * unitCount;
			BufferSize = bytesPerFrame * BufferFrames;
			BufferCount = unitCount * 4;
			MuteData = new byte[BufferSize];
			WaveFormatEx = new WAVEFORMATEX()
			{
				wFormatTag = FormatTag,
				nChannels = Channels,
				nSamplesPerSec = (uint)sampleRate,
				nAvgBytesPerSec = (uint)(bytesPerFrame * sampleRate),
				nBlockAlign = (ushort)bytesPerFrame,
				wBitsPerSample = Bits,
				cbSize = 0
			};
			WaveHeaderPtrs = new IntPtr[BufferCount];
			DeviceOpened = new AutoResetEvent(false);
			DeviceClosed = new AutoResetEvent(false);
			CallbackEnabled = new AutoResetEvent(false);
			BufferReady = new AutoResetEvent(false);
			TotalWaveHeader = Marshal.AllocHGlobal(Marshal.SizeOf<WAVEHDR>() * BufferCount);
			TotalBuffer = Marshal.AllocHGlobal(BufferSize * BufferCount);
		}

		~Wave() {
			Free();
		}

		private void Free() {
			DeviceOpened?.Dispose();
			DeviceClosed?.Dispose();
			CallbackEnabled?.Dispose();
			BufferReady?.Dispose();
			if (IntPtr.Zero != TotalWaveHeader) {
				Marshal.FreeHGlobal(TotalWaveHeader);
				TotalWaveHeader = IntPtr.Zero;
			}
			if (IntPtr.Zero != TotalBuffer) {
				Marshal.FreeHGlobal(TotalBuffer);
				TotalBuffer = IntPtr.Zero;
			}
		}

		private void ResetHeader() {
			for (int i = 0; i < BufferCount; ++i) {
				var header = new WAVEHDR()
				{
					dwBufferLength = (uint)BufferSize,
					lpData = IntPtr.Add(TotalBuffer, BufferSize * i),
				};
				WaveHeaderPtrs[i] = IntPtr.Add(TotalWaveHeader, Marshal.SizeOf<WAVEHDR>() * i);
				Marshal.Copy(MuteData, 0, header.lpData, BufferSize);
				Marshal.StructureToPtr(header, WaveHeaderPtrs[i], false);
			}
		}

		private void ClearStatus() {
			DeviceOpened.Reset();
			DeviceClosed.Reset();
			CallbackEnabled.Reset();
			BufferReady.Reset();
			NotifyClose = false;
			NotifyEndOfFile = false;
			NotifyMute = false;
			Muted = true;
		}

		protected void OpenDevice() {
			CloseDevice();
			BufferThread = new Thread(() => {
				ResetHeader();
				ClearStatus();
				if (InitializeTask()) {
					BufferTask();
					FinalizeTask();
				}
				ClearStatus();
				DeviceHandle = IntPtr.Zero;
			})
			{
				Priority = ThreadPriority.Highest,
				IsBackground = true
			};
			BufferThread.Start();
		}

		protected void CloseDevice() {
			if (IntPtr.Zero == DeviceHandle) {
				return;
			}
			NotifyClose = true;
			BufferThread?.Join();
		}

		protected abstract bool InitializeTask();

		protected abstract void FinalizeTask();

		protected abstract void BufferTask();

		public void Dispose() {
			CloseDevice();
			Free();
			GC.SuppressFinalize(this);
		}

		public void SetDevice(uint deviceId) {
			CloseDevice();
			DeviceId = deviceId;
			OpenDevice();
		}

		public void Start() {
			NotifyMute = false;
			Muted = false;
		}

		public void Stop() {
			NotifyMute = true;
			for (int i = 0; i < 100 && !Muted; i++) {
				Thread.Sleep(10);
			}
			NotifyMute = false;
		}
	}
}

File: WinAPI/Wave.cpp
#include "Wave.h"

#pragma comment(lib, "winmm.lib")

Wave::Wave(int32_t sampleRate, int32_t channels, EBufferType bufferType, int32_t bufferSamples, int32_t bufferCount) {
	auto isFloat = ((uint16_t)bufferType & (uint16_t)EBufferType::FLOAT) != 0;
	auto bits = (uint16_t)bufferType & (uint16_t)EBufferType::BITS_MASK;
	auto bytesPerSample = channels * bits >> 3;
	SampleRate = sampleRate;
	Channels = channels;
	DeviceId = WAVE_MAPPER;
	BufferType = bufferType;
	BufferSamples = bufferSamples;
	BufferSize = bufferSamples * bytesPerSample;
	BufferCount = bufferCount;
	WaveFormatEx = {};
	WaveFormatEx.wFormatTag = isFloat ? 3 : 1;
	WaveFormatEx.nChannels = (WORD)channels;
	WaveFormatEx.nSamplesPerSec = (DWORD)sampleRate;
	WaveFormatEx.wBitsPerSample = (WORD)bits;
	WaveFormatEx.nBlockAlign = (WORD)((bits >> 3) * (WORD)channels);
	WaveFormatEx.nAvgBytesPerSec = WaveFormatEx.nBlockAlign * (DWORD)sampleRate;
	WaveFormatEx.cbSize = 0;
}

Wave::~Wave() {
	CloseDevice();
}

void Wave::SetDevice(uint32_t deviceId, bool withStart) {
	auto enable = DeviceEnabled;
	CloseDevice();
	DeviceId = deviceId;
	if (enable || withStart) {
		OpenDevice();
	}
}

void Wave::Start() {
	Pause = false;
	Paused = false;
	Playing = DeviceEnabled;
}

void Wave::Stop() {
	Pause = true;
	if (Playing) {
		WaitEnable(&Paused);
		Playing = false;
	}
}

void Wave::OpenDevice() {
	CloseDevice();
	ThreadBuffer = CreateThread(
		nullptr,
		0,
		(LPTHREAD_START_ROUTINE)Task,
		this,
		0,
		(LPDWORD)&ThreadBuffer
	);
	if (nullptr == ThreadBuffer) {
		return;
	}
	SetThreadPriority(ThreadBuffer, THREAD_PRIORITY_HIGHEST);
}

void Wave::CloseDevice() {
	if (nullptr == DeviceHandle) {
		return;
	}
	Closing = true;
	WaitForSingleObject(ThreadBuffer, 2000);
}

void Wave::WaitEnable(bool *flag) {
	for (int i = 0; i < 200 && !*flag; i++) {
		Sleep(10);
	}
}

void Wave::WaitDisable(bool *flag) {
	for (int i = 0; i < 200 && *flag; i++) {
		Sleep(10);
	}
}

void Wave::AllocateHeader() {
	WaveHeaders = (WAVEHDR*)malloc(sizeof(WAVEHDR) * BufferCount);
	if (nullptr == WaveHeaders) {
		return;
	}
	memset(WaveHeaders, 0, sizeof(WAVEHDR) * BufferCount);
	MuteData = malloc(BufferSize);
	if (nullptr != MuteData) {
		if (BufferType == EBufferType::INT8) {
			memset(MuteData, 128, BufferSize);
		} else {
			memset(MuteData, 0, BufferSize);
		}
	}
	for (int i = 0; i < BufferCount; ++i) {
		WAVEHDR header{};
		header.lpData = (LPSTR)malloc(BufferSize);
		header.dwBufferLength = (uint32_t)BufferSize;
		header.dwFlags = WHDR_BEGINLOOP | WHDR_ENDLOOP;
		memcpy_s(header.lpData, BufferSize, MuteData, BufferSize);
		memcpy_s(WaveHeaders + i, sizeof(WAVEHDR), &header, sizeof(WAVEHDR));
	}
}

void Wave::DisposeHeader() {
	if (nullptr == WaveHeaders) {
		return;
	}
	for (int i = 0; i < BufferCount; ++i) {
		auto pHeader = WaveHeaders + i;
		if (nullptr == pHeader) {
			continue;
		}
		if (nullptr != pHeader->lpData) {
			free(pHeader->lpData);
		}
	}
	free(WaveHeaders);
}

void Wave::ClearFlags() {
	Closing = false;
	Paused = false;
	DeviceEnabled = false;
	CallbackEnabled = false;
}

DWORD Wave::Task(LPVOID *lpParam) {
	auto self = (Wave*)lpParam;
	InitializeCriticalSection(&self->LockBuffer);
	self->AllocateHeader();
	self->ClearFlags();
	if (self->InitializeTask()) {
		self->BufferTask();
		self->FinalizeTask();
	}
	self->DisposeHeader();
	self->ClearFlags();
	self->DeviceHandle = nullptr;
	if (nullptr != self->LockBuffer.DebugInfo) {
		DeleteCriticalSection(&self->LockBuffer);
		self->LockBuffer.DebugInfo = nullptr;
	}
	return 0;
}
